ECDSA Alg improvements

Spomky opened the pull request:
* `fgrosse/phpasn1` dependency removed
* New EC signature converter
* Asn1/PointsManipulator removed as useless

Hi @lcobucci,

As discussed, here is a proposal to get ride of `fgrosse/phpasn1` and fix trouble with that library.
This could be a first step to fix #259 by backporting this solution.

The class `ECSignature` may need to be improved as well.
Feel free to propose correction on that class if so.

lcobucci commented:
@Spomky thanks a lot for your contribution, it's quite nice to remove that dependency.

The only thing is that I'd prefer to keep the internal interface and the signature converter as an implementation of it. The main motivation is to isolate things and allow us to abstract that bit from the rest of the lib.

If you prefer, I can take it from here :+1:

Spomky commented:
I updated the PR and restored the PointsManipulator interface.
The class I added now implements that interface.

I will also take time to add more tests for these algorithms.

lcobucci commented:
@Spomky thank you very much

Spomky commented:
Hi @lcobucci ,

I added some tests taken from the RFC6979.
The one from the [RFC7520](https://tools.ietf.org/html/rfc7520#section-4.3) cannot be added because the payload is not a JSON object and parsing exceptions are thrown (you may have an idea to integrate it).
Anyway it works fine now and the algorithms are slightly faster (approx. -25% of execution time). The only problem is that some mutations are not detected and I am not sure it will be easy to get ride of these.

Let me know if you have ideas to fix these metrics. I will modify my PR accordingly

Spomky commented:
Great. Thank you for the review.
I will take care of the modifications.

Spomky commented on src/Signer/Ecdsa/ECSignature.php:
```diff
@@ -0,0 +1,99 @@
+ * Copyright (c) 2014-2018 Spomky-Labs
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
```
No problem :+1:

Spomky commented on src/Signer/Ecdsa/ECSignature.php:
```diff
@@ -0,0 +1,99 @@
+
+        return pack(
+            'H*',
+            '30' . ($Rl + $Sl + 4 > 128 ? '81' : '') . dechex($Rl + $Sl + 4)
```
Ok I will add private constants (e.g. `ASN_SEQUENCE_OBJECT`, `ASN_INTEGER_OBJECT`)

Spomky commented on src/Signer/Ecdsa/ECSignature.php:
```diff
@@ -0,0 +1,99 @@
+    {
+        $hex = unpack('H*', $signature)[1];
+        if (mb_substr($hex, 0, 2, '8bit') !== '30') { // SEQUENCE
+            throw new RuntimeException();
```
You are right, all these exception should have a clear message.

Spomky commented on test/functional/EcdsaTokenTest.php:
```diff
@@ -355,4 +359,125 @@ public function everythingShouldWorkWhenUsingATokenGeneratedByOtherLibs(): void
+     * @covers \Lcobucci\JWT\Signer\Ecdsa\Sha384
+     * @covers \Lcobucci\JWT\Signer\Ecdsa\Sha512
+     */
+    public function theVectorsFromRFC6978CanBeVerified(Ecdsa $signer, Key $key, string $payload, string $expected): void
```
Ok. These tests will have their own testing class

Spomky commented on test/functional/EcdsaTokenTest.php:
```diff
@@ -355,4 +359,125 @@ public function everythingShouldWorkWhenUsingATokenGeneratedByOtherLibs(): void
+     */
+    public function dataRFC6979(): array
+    {
+        return [
```
I love that. This will be done accordingly.

Spomky commented:
Hi @lcobucci,

I modified my PR few hours ago.
New deprecation notices from PHPUnit are catched now. Not sure it is important to fix these errors right now.
I am playing with my `ECSignature` and trying to improve it a bit. Just let me few more days to get it working correctly.

lcobucci commented:
@Spomky thanks a lot! I'm actually addressing those deprecations right now, so don't worry about them. Once I'm done you will be able to rebase your branch and sync it with `master` and things will be alright :+1:

Spomky commented:
Hi @lcobucci 

I hope everything is going well for you.
All the best for this new year 2019!

This PR looks good to me. I don't see any optimization at the moment.

lcobucci commented:
@Spomky alright I wasn't sure if you had finished. Thank you for your contribution, I'd take a look on it this week and get it merged in :+1: 

All the best for you too!
